File: trace_callstack.py
"""
trace_callstack.py — 路线 Alpha
ReadFile SSKF 触发 → 读 Thread.backtrace → 定位 AcquireSMD 调用者
"""
import frida
import subprocess
import sys
import os
import time
from datetime import datetime

DUMP_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'apollo_dump')

def find_pid():
    r = subprocess.run(['tasklist','/FI','IMAGENAME eq FreeStyle.exe','/FO','CSV','/NH'],
                       capture_output=True, text=True)
    for line in r.stdout.strip().split('\n'):
        if 'freestyle' in line.lower():
            parts = line.split(',')
            if len(parts) >= 2:
                return int(parts[1].strip('"'))
    return None

def main():
    pid = find_pid()
    if not pid:
        print("[ERROR] FreeStyle.exe not found. Start the game first.")
        return

    ts = datetime.now().strftime('%Y%m%d_%H%M%S')
    log_path = os.path.join(DUMP_DIR, f'callstack_{ts}.txt')
    os.makedirs(DUMP_DIR, exist_ok=True)
    log = open(log_path, 'w', encoding='utf-8')
    print(f"Log: {log_path}")

    def log_msg(msg):
        print(msg)
        log.write(msg + '\n')
        log.flush()

    session = frida.attach(pid)
    log_msg(f"Attaching PID {pid}...")

    with open(os.path.join(os.path.dirname(__file__), 'trace_callstack.js'), 'r', encoding='utf-8') as f:
        js_code = f.read()

    script = session.create_script(js_code)

    def on_msg(msg, data):
        if msg['type'] == 'error':
            log_msg(f"  [JS ERROR] {msg.get('description', msg)}")
            return
        payload = msg.get('payload', {})
        t = payload.get('t', '')
        if t == 'SSKF':
            name = payload.get('name', '?')
            sz = payload.get('sz', 0)
            stack = payload.get('stack', [])
            log_msg(f"\n=== SSKF: {name} ({sz} bytes) ===")
            for i, frame in enumerate(stack):
                marker = ' <--' if 'FreeStyle.exe' in frame else ''
                log_msg(f"  #{i}: {frame}{marker}")
        elif t == 'diag':
            log_msg(f"  [diag] {payload.get('msg', '')}")
        elif t == 'ready':
            log_msg(f"  {payload.get('msg', '')}")
        else:
            log_msg(f"  {payload}")

    script.on('message', on_msg)
    script.load()

    session.on('detached', lambda reason, crash:
        log_msg(f"\n!!! PROCESS EXIT: reason={reason}, crash={crash} !!!"))

    log_msg("\n=== SSKF + Call Stack Tracer ===")
    log_msg("操作: 在游戏里动一下, 触发 SMD 加载")
    log_msg("Ctrl+C 退出")

    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        log_msg("\nDetaching...")
    finally:
        session.detach()
        log.close()
        print(f"\nDone. Log: {log_path}")

if __name__ == '__main__':
    main()

File: trace_callstack.js
// ============================================================
// trace_callstack.js — 路线 Alpha
// ReadFile SSKF 检测 + Thread.backtrace → 定位 AcquireSMD 入口
// ============================================================

var kernel32 = Process.getModuleByName('kernel32.dll');
var ReadFileAddr = kernel32.getExportByName('ReadFile');
send({t: 'diag', msg: 'kernel32=' + kernel32.base + ' ReadFile=' + ReadFileAddr});

Interceptor.attach(ReadFileAddr, {
    onEnter: function(args) {
        this.hFile = args[0];
        this.lpBuffer = args[1];
        this.nBytes = args[2].toInt32();
    },
    onLeave: function(retval) {
        if (this.nBytes < 32 || retval.toInt32() === 0) return;

        try {
            var buf = this.lpBuffer.readByteArray(Math.min(this.nBytes, 64));
            var u8 = new Uint8Array(buf);
        } catch(e) { return; }

        // SSKF magic check
        if (u8[0] !== 0x53 || u8[1] !== 0x53 || u8[2] !== 0x4B || u8[3] !== 0x46) return;

        // Extract filename from SSKF header
        var fname = '';
        for (var i = 8; i < Math.min(this.nBytes, 64); i++) {
            if (u8[i] === 0) break;
            fname += String.fromCharCode(u8[i]);
        }

        // === CALL STACK ===
        var stackFrames = [];
        try {
            var bt = Thread.backtrace(this.context, Backtracer.ACCURATE);
            for (var j = 0; j < Math.min(bt.length, 20); j++) {
                var addr = bt[j];
                var mod = Process.findModuleByAddress(addr);
                if (mod) {
                    var off = addr.sub(mod.base);
                    stackFrames.push(mod.name + '+' + off.toString(16));
                } else {
                    stackFrames.push(addr.toString());
                }
            }
        } catch(e) {
            stackFrames.push('backtrace error: ' + e);
        }

        send({
            t: 'SSKF',
            name: fname,
            sz: this.nBytes,
            stack: stackFrames
        });
    }
});

send({t: 'ready', msg: 'SSKF callstack tracer active'});